arches/app/views/workflow_history.py

```diff
@@ -0,0 +1,75 @@
+import json
+
+from django.db import transaction
+from django.db.models import F, JSONField, Value
+from django.db.models.expressions import CombinedExpression
+from django.db.utils import IntegrityError
+from django.utils.translation import gettext as _
+from django.views.generic import View
+
+from arches.app.utils.permission_backend import user_is_resource_reviewer
+from arches.app.utils.response import JSONErrorResponse, JSONResponse
+from arches.app.models import models
+
+
+class WorkflowHistoryView(View):
+
+    def get(self, request, workflowid):
+        if not user_is_resource_reviewer(request.user):
+            return JSONErrorResponse(_("Request Failed"), _("Permission Denied"), status=403)
+        try:
+            workflow_history = models.WorkflowHistory.objects.get(workflowid=workflowid, user=request.user)
+        except models.WorkflowHistory.DoesNotExist:
+            workflow_history = {}
+        return JSONResponse(workflow_history, status=200)
+
+    def post(self, request, workflowid):
+        if not user_is_resource_reviewer(request.user):
+            return JSONErrorResponse(_("Request Failed"), _("Permission Denied"), status=403)
+
+        data = json.loads(request.body)
+        stepdata = data.get("stepdata", {})
+        componentdata = data.get("componentdata", {})
+
+        # TODO(Django 5.0) rewrite as a simpler update_or_create()
+        # call using different `defaults` vs. `create_defaults`
```

Comment:
`update_or_create()` almost does what we need, but you can't use F expressions to reference the current state of a field inside a create statement (makes sense!).

I need the current state of the workflow fields to combine the keys. In Django 5.0 you can provide different behavior for create and update, so we could slim this down a bit. Right now this is just a riff on how `update_or_create()` works.

See [Django forum](https://forum.djangoproject.com/t/feature-idea-update-or-create-to-allow-different-defaults-for-create-and-update-operations/18300)